### PR TITLE
April 2026 rebase, part 3

### LINKED ISSUE #360: Bug: Malformed leap month suffix while implementing Chinese calendar
**Browser:** Safari 26.4

**Polyfill cdn:** https://esm.sh/@js-temporal/polyfill

**Issue:** I'm leveraging this polyfill for a chinese calendar I am implementing  to also include the leap year functionality from `Temporal`. The error I came across was `Error: Unexpected leap month suffix: Mo3`. 

To recreate:
```
const today = (Temporal.Now.plainDateISO()).toString();
const date = Temporal.PlainDate.from(today);
const chineseDate = date.withCalendar('chinese');

// Calling chineseDate.year, chineseDate.month, chinese.monthCode, chineseDate.monthsInYear recreates the error
```

It seems like a malformed month code issue stemming from https://github.com/js-temporal/temporal-polyfill/blob/ef32041bd094af68f6ca3e030ddf594feb873451/lib/calendar.ts#L777

Any help or direction would be much appreciated!

**Screenshots of the error:**

Screenshot 2026-04-21 at 10 47 38 PM

Screenshot 2026-04-21 at 10 45 32 PM

### PR DESCRIPTION
**Pull request by ptomato**
This batch of commits mainly deals with working around weirdnesses in newer ICU versions, and bugfixes to pass the test262 tests dealing with calendar calculations.

There is one normative change, from the November 2025 TC39 plenary. It deals with rounding durations in edge cases and is unlikely to break any working code.

Closes: #360

### CODE REVIEW

**Comment by craig-o-curtis** [comment]
The normative change in this PR (`50d66d2`, porting proposal-temporal `5dd0b0d97ee1` / tc39/proposal-temporal#3168) fixes a user-visible wrong answer on `main` today. Sharing the evidence in case it helps prioritise the review, since the PR description understandably describes it as "unlikely to break any working code".

```js
const from = Temporal.PlainDateTime.from("2024-01-31T00:00:00");
const to   = Temporal.PlainDateTime.from("2024-02-29T12:00:00");

from.until(to, { largestUnit: "month" }).toString();
// "P29DT12H" on both — correct, and the reason r1 is 0 months

from.until(to, { largestUnit: "month", smallestUnit: "month", roundingMode: "trunc" }).toString();
// main: "PT0S"                after 50d66d2: "P1M"

from.until(to, { largestUnit: "month" }).total({ unit: "month", relativeTo: from });
// main: 1.0172413793103448    after 50d66d2: 1.0161290322580645

Temporal.Duration.from("P29DT12H").round({
  smallestUnit: "month", roundingMode: "floor",
  relativeTo: Temporal.PlainDate.from("2024-01-31"),
}).toString();
// main: "PT0S"                after 50d66d2: "P1M"
```

Both columns are measured here, from two `NODE_ENV=production` builds — `main` at `c8f344c`, and the same tree with `50d66d2` applied on top. The post-fix values agree with Chromium 153.0.8010.12; on #3168 ptomato recorded that SpiderMonkey, temporal_rs and the FullCalendar polyfill agree with V8 too.

`Jan 31 + 1 month` is `Feb 29` only by constraining, so `until` doesn't count it as a month and `r1` is 0. The first window is then `Jan 31 → Feb 29`, which ends before the `Feb 29T12:00` target, so `progress` exceeds 1 and the result is taken against bounds that don't contain it. In production builds the `${unit} was 0 days long` assert is stripped, so it returns a wrong value instead of throwing. Only a `relativeTo` on the 29th, 30th or 31st can reach it.

Scanning Chromium 153.0.8010.12 against published 0.5.1: `total` 16 mismatches in 2,016 rows, `round` 15 in 12,960, `until`/`since` with a calendar `smallestUnit` 15 in 9,940 — all `unit: "month"`, and only in the directed rounding modes (`ceil`/`floor`/`trunc`). The half modes agree either way, because with `progress > 1` they still select `r2`. `until` without rounding agrees on every row, and `Duration.compare` reaches no nudge window at all (256 rows, 0 mismatches).

One note for whoever reviews it: #3168's own reproducer (`new Temporal.Duration(1, 0, 0, 0, 1)` relative to `2020-02-29`, unit `years`) **passes** on 0.5.1 — there `r1` is already 1 from the duration's own years field, so no retry is needed. The month cases above are what expose it, and test262 doesn't cover this yet.

Verification of `50d66d2` applied alone on top of `main`, Node 24.21.0 / ICU 78.3:

- `npm test` — 607 passed, 4 suites, 0 failed
- `npm run test262` (pinned `ac3035f0`, production build) — 4855 finished, 4828 passed, **0 failed, 0 passed unexpectedly**, 27 expected failures, 6 missing
- `npm run lint` — clean

It applies cleanly on `main` by itself and touches only `lib/ecmascript.ts`, so if it would help to have this one commit split out as a small standalone PR while the rest of the rebase waits for review, I'm happy to open it — just say the word. Otherwise I'll leave it here where it belongs.

Found while testing [`@northguild/gmt`](https://github.com/northguild/gmt) against the specification; it carries a dormant workaround until a release lands.

**Review by justingrant — APPROVED** [review]
Looks great, thanks for porting these commits, we're getting close to present day!

Copilot had some feedback you may want to look at, but I didn't see anything obvious to worry about.

File: lib/intl.ts [hunk 1]
```diff
@@ -462,7 +464,7 @@ function yearMonthAmend(originalOptions: OptionsType<Temporal.PlainYearMonth>) {
     delete options.dateStyle;
     Object.assign(options, dateStyleHacks[style]);
   }
-  if (!('year' in options || 'month' in options || 'era' in options)) {
+  if (!('year' in options || 'month' in options)) {
     if (hasAnyDateTimeOptions(originalOptions)) {
       throw new TypeError(`cannot format PlainYearMonth with options [${Object.keys(originalOptions)}]`);
     }
```
[the remark this answers is not in the document]
Comment (ptomato) [reply lib/intl.ts:470]: This is intentional, per https://github.com/tc39/proposal-temporal/issues/3049 and the discussion there.

File: lib/intl.ts [hunk 2]
```diff
@@ -577,14 +579,7 @@ function instantAmend(optionsParam: OptionsType<Temporal.Instant>) {
...
-  return (
-    'year' in options ||
-    'month' in options ||
-    'day' in options ||
-    'weekday' in options ||
-    'dateStyle' in options ||
-    'era' in options
-  );
+  return 'year' in options || 'month' in options || 'day' in options || 'weekday' in options || 'dateStyle' in options;
 }
```
[the remark this answers is not in the document]
Comment (ptomato) [reply lib/intl.ts:583]: Same as above.
